Template exceptions fix in database and file importers

Auriga2 opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Make sure your branch name is prefixed by `feature`, `bugfix` or `release`
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Prerequisites

- [x] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change
Templates in database and file importers now handle some edge cases and prevent exceptions, specifically when the templates are of an incorrect format. This was specially the case with legacy templates. More notes and discussions are in https://github.com/constellation-app/constellation/issues/1652 after the comment by [GammaVel](https://github.com/GammaVel)  
<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs
n/a
<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?
Templates are more usable
<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an existing file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->

### Benefits
Better UX
<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks
n/a
<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process
As described in detain in https://github.com/constellation-app/constellation/issues/1652 Legacy/incorrect format could be manually created by editing a new template as in the screen captured there. Test both importers with all combinations of options, paying attention to the attributes section. Verify they are updated correctly as per the options saved in the loaded template.

<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g. buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
Related to https://github.com/constellation-app/constellation/issues/1652

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedIO.java:
```diff
@@ -225,6 +227,8 @@ public static void loadParameters(final Window parentWindow, final DelimitedImpo
 
     private static void loadParameterFile(final DelimitedImportController importController, final File delimIoDir,
             final String templName) {
+        List<String> missingUserAttributes = new ArrayList<>();
```
final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/delimited/ImportDelimitedIO.java:
```diff
@@ -303,18 +317,32 @@ private static void loadParameterFile(final DelimitedImportController importCont
                 final String message = String.format("Can't find schema factory '%s'", destination);
                 NotifyDisplayer.displayAlert(LOAD_TEMPLATE, "Destination Schema Error", message, Alert.AlertType.ERROR);
             }
+        } catch (final NullPointerException ex) {
```
I think my preference would be to do null checks over catching a null pointer as it's better practice that way. Sonar will also be flagging this. Potentially what you could is do a null check and then throw a different kind of exception (probably runtime) in the else block

Auriga2 replied:
I tried to avoid checking null in ~17 lines which are unlikely/less likely to occur. NullPointerException is thrown at the JsonNode.get() hence cannot throw another exception and handle it in this code.

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/ImportJDBCIO.java:
```diff
@@ -273,19 +292,32 @@ private static void loadParameterFile(final JDBCImportController importControlle
+                final String message = String.format("Can't find schema factory '%s'", destination);
+                NotifyDisplayer.displayAlert(LOAD_TEMPLATE, "Destination Schema Error", message, Alert.AlertType.ERROR);
             }
+        } catch (final NullPointerException ex) {
```
similar here

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/jdbc/ImportJDBCIO.java:
```diff
@@ -211,6 +218,8 @@ public static void loadParameters(final Window parentWindow, final JDBCImportCon
     }
 
     private static void loadParameterFile(final JDBCImportController importController, final File delimIoDir, final String templName) {
+        List<String> missingUserAttributes = new ArrayList<>();
```
final

Auriga2 commented:
@antares1470 I think I fixed it as you mentioned. Also refactored the class by moving the common code to a utility class. (Ideally the `saveParameters` could also be refactored to reduce the Cognitive Complexity)

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java:
```diff
@@ -0,0 +1,167 @@
+    public static List<ImportDefinition> getImportDefinitions(final ImportController importController, final JsonNode root, final String templName) {
+        final List<ImportDefinition> definitions = new ArrayList<>();
+        int tabCount = 0;
+        final JsonNode source = TemplateUtilities.getRequiredFieldFromTemplate(root, SOURCE);
```
you should be able to call the function without the explicit call of `TemplateUtilities` since you are still operating in that class. Similarly for 2 of the next 3 lines after this one

Auriga2 replied:
Oh initially added in ImportDelimitedIO.java. A copy paste issue! Fixed now.

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/TemplateUtilities.java:
```diff
@@ -0,0 +1,167 @@
+            final JsonNode filterNode = TemplateUtilities.getRequiredFieldFromTemplate(definitionNode, FILTER);
+            final String script = TemplateUtilities.getRequiredFieldFromTemplate(filterNode, SCRIPT).textValue();
+            final JsonNode columnsArray = filterNode.withArray(COLUMNS);
+            final ArrayList<String> columns = new ArrayList<>();
```
Can you make the left hand side just a `List<String>`

Auriga2 replied:
Just copy pasted, fixed now.

mimosa2 commented:
@Auriga2, A minor update - include author & class details for TemplateUtilities.java.
